notif.go

```diff
@@ -33,34 +29,57 @@ func (nn *netNotifiee) Connected(n inet.Network, v inet.Conn) {
 	default:
 	}
 
+	dht.plk.Lock()
+	defer dht.plk.Unlock()
+
+	conn, ok := nn.peers[v.RemotePeer()]
+	if ok {
+		conn.refcount += 1
+		return
+	}
+
+	ctx, cancel := context.WithCancel(dht.Context())
+
+	nn.peers[v.RemotePeer()] = &peerTracker{
+		refcount: 1,
+		cancel:   cancel,
+	}
+
 	go func() {
 
 		// Note: We *could* just check the peerstore to see if the remote side supports the dht
 		// protocol, but its not clear that that information will make it into the peerstore
 		// by the time this notification is sent. So just to be very careful, we brute force this
 		// and open a new stream
 
-		// TODO: There's a race condition here where the connection may
-		// not be open (and we may sit here trying to connect). I've
-		// added a timeout but that's not really the correct fix.
-
-		ctx, cancel := context.WithTimeout(dht.Context(), dhtCheckTimeout)
-		defer cancel()
-		s, err := dht.host.NewStream(ctx, v.RemotePeer(), ProtocolDHT, ProtocolDHTOld)
-
-		switch err {
-		case nil:
-			s.Close()
-			// connected fine? full dht node
-			dht.Update(dht.Context(), v.RemotePeer())
-		case mstream.ErrNotSupported:
-			// Client mode only, don't bother adding them to our routing table
-		case io.EOF:
-			// This is kindof an error, but it happens someone often so make it a warning
-			log.Warningf("checking dht client type: %s", err)
-		default:
-			// real error? thats odd
-			log.Errorf("checking dht client type: %s", err)
+		for {
+			s, err := dht.host.NewStream(ctx, v.RemotePeer(), ProtocolDHT, ProtocolDHTOld)
+
+			// Canceled.
+			if ctx.Err() != nil {
```

Comment:
isn't it possible that err is nil and ctx.Err() is not nil because of a race? 
In this case we would leak s.